barretenberg/cpp/src/barretenberg/transcript/README.md

```diff
@@ -0,0 +1,853 @@
+    previous_challenge:            // For duplex sponge c_next = H(c_prev || data)
+
+    // Security-critical state is private, accessed via extract_transcript_tag()
+```
+
+---
+
+## Usage Examples
+
+### Standard Prover/Verifier Pattern
+
+```cpp
+// === PROVER ===
+auto prover_transcript = std::make_shared<NativeTranscript>();
+
+// Round 1: Send wire commitments
+for (auto& comm : wire_commitments) {
+    prover_transcript->send_to_verifier("wire_comm", comm);
+}
+
+// Round 2: Receive challenge, send derived commitments
+auto [beta, gamma] = prover_transcript->get_challenges<FF>({"beta", "gamma"});
+compute_grand_product_polynomials(beta, gamma);
+for (auto& z_comm : z_commitments) {
+    prover_transcript->send_to_verifier("z_comm", z_comm);
+}
+
+auto proof = prover_transcript->export_proof();
+
+// === VERIFIER ===
+auto verifier_transcript = std::make_shared<NativeTranscript>();
+verifier_transcript->load_proof(proof);
+
+// Mirror prover's exact sequence
+for (size_t i = 0; i < num_wires; i++) {
+    auto comm = verifier_transcript->receive_from_prover<Commitment>("wire_comm");
+    // Verifier stores commitments...
+}
+
+auto [beta, gamma] = verifier_transcript->get_challenges<FF>({"beta", "gamma"});
+
+for (size_t i = 0; i < num_relations; i++) {
+    auto z_comm = verifier_transcript->receive_from_prover<Commitment>("z_comm");
+    // Verifier stores z commitments...
+}
+```
+
+### Handling Public Inputs
+
+Public inputs are typically **not** in the proof (verifier reconstructs them):
+
+```cpp
+// PROVER: Hash but don't add to proof
+for (auto& input : public_inputs) {
+    transcript->add_to_hash_buffer("public_input", input);
+}
+
+// VERIFIER: Hash the reconstructed inputs
+std::vector<FF> reconstructed_inputs = get_public_inputs_from_context();
+for (auto& input : reconstructed_inputs) {
+    transcript->add_to_hash_buffer("public_input", input);
+}
+```
+
+
+### Shared Transcript Between Prover Stages
+
+```cpp
+// Stage 1: OINK prover
+auto transcript = std::make_shared<NativeTranscript>();
+OinkProver oink_prover(proving_key, transcript);
+oink_prover.prove();
+auto oink_proof = transcript->export_proof();  // Extract OINK portion
+
+// Stage 2: Decider prover (continues on same transcript)
+DeciderProver decider_prover(proving_key, transcript);
+decider_prover.prove();
+auto decider_proof = transcript->export_proof();  // Extract Decider portion
+
+// Combined proof: concatenate [oink_proof || decider_proof]
+```
+
+---
+
+## Best Practices
+
+### 1. **In-Circuit: Unset Free Witness Tags Carefully**
+
+In recursive verification, you may need to bypass free witness checks:
+
+```cpp
+// Only do this if you're CERTAIN the value is safe
+commitment.unset_free_witness_tag();
+
+// Better: Ensure values come from transcript with proper origin tags
+auto commitment = transcript->receive_from_prover<Commitment>("comm");
+// No need to unset - it has a proper origin tag
+```
+
+**When unsetting is needed**:
+- VK commitments
```

Comment:
I don't understand why it's needed for vk commitments.
Also, maybe add a sentence or two on what unsetting does.